Dip/Dip/Dip.swift

```diff
@@ -253,14 +263,44 @@ public class DependencyContainer {
   ///Pool to hold instances, created during call to `resolve()`. 
   ///Before `resolve()` returns pool is drained.
   class ResolvedInstances {
-    var resolvedInstances = [DefinitionKey: Any]()
+    var resolvedInstances = [String: Any]()
 
     func storeResolvedInstance<T>(instance: T, forKey key: DefinitionKey?) {
-      self.resolvedInstances[key] = instance
+      self[key] = instance
     }
     
     func previouslyResolved<T, F>(key: DefinitionKey?, definition: DefinitionOf<T, F>) -> T? {
-      return (definition.resolvedInstance ?? self.resolvedInstances[key]) as? T
+      return (definition.resolvedInstance ?? self[key]) as? T
+    }
+    
+    subscript(key: DefinitionKey?) -> Any? {
+      get {
+        guard let key = resolvedKey(key) else { return nil }
+        return resolvedInstances[key]
+      }
+      set {
+        guard let key = resolvedKey(key) else { return }
+        resolvedInstances[key] = newValue
+      }
+    }
+    
+    func resolvedKey(key: DefinitionKey?) -> String? {
+      guard let key = key else { return nil }
+      
+      let stringKey: String
+      switch key.associatedTag {
+      case .String(let tagValue)? where tagValue.hasPrefix("InjectedWeak<"):
```

Comment:
Couldn't we do more secure and error-proof tests rather than string-based comparisons here?